### PR TITLE
feat(extensions): add VaultPermissionManager actor-authorization layer

### PR DESCRIPTION
**Pull request by tsu-j**
The `VaultPermissionManager` extension lets a BAP-578 NFT owner grant a third-party operator the right to drive their agent's `handleAction` entry point, within a time-bounded permission tier. The operator never holds the NFT and cannot transfer, list, or re-logic it.

It complements the existing `IPolicyGuard` framework. `IPolicyGuard` answers "is this action allowed?" (per-call). `IVaultPermissionManager` answers "is this actor authorized to drive the agent at all?" (per-actor). A full agent stack uses both.

### Ticket

N/A

#### Summary of Changes

Adds an interface, a UUPS reference implementation, two test mocks, and 30 tests, in the style of the existing `extensions/IPolicy*` family.

- `contracts/extensions/IVaultPermissionManager.sol`: interface. `PermissionLevel` tiers (NONE/READ/WRITE/ADMIN), `createVault`, `grantPermission`, `revokePermission`, `forwardHandleAction`, `checkPermission`, `canForward`, `vaultExists`.
- `contracts/extensions/VaultPermissionManagerReference.sol`: UUPS reference implementation. Per-agent vault namespaces, time-bounded grants, owner-or-admin grant authority, `nonReentrant` forwarding, and auto-revoke on NFT transfer (a grant stamps `ownerOf(tokenId)` at creation and is treated as revoked once the NFT changes hands).
- `contracts/mocks/MockBAP578ForVPM.sol`, `contracts/mocks/MockLogicModule.sol`: minimal mocks for the unit tests.
- `test/VaultPermissionManager.test.js`: 30 tests covering vault lifecycle, grant/revoke including the admin path, the forward path, cross-grantee/vault/token isolation, re-grant semantics, init plus upgrade auth, and auto-revoke including the documented transfer-and-back edge case.

Targets `pragma solidity ^0.8.28` and `@openzeppelin/contracts-upgradeable@^4.9.x` to match the repo. All 30 tests pass under the repo toolchain with no compiler warnings.

#### Notes

Composition with `IPolicyGuard`:

1. External caller invokes `VPM.forwardHandleAction(tokenId, vaultId, action, data)`.
2. VPM checks `canForward` (WRITE grant, not expired).
3. VPM dispatches into the bound logic module's `handleAction`.
4. The logic module gates its own outbound calls through `IPolicyGuard.validate(...)`.
5. After a successful action, the logic calls `IPolicyGuard.commit(...)`.

A logic module that wants to be drivable through VPM only needs the existing `handleAction` shape, with no new interface on its side.

Documented edge case: if an NFT returns to the original granter without an explicit revoke in between, stale grants reactivate (the `ownerAtGrant` stamp matches again). The test `reactivates a stale grant if the NFT returns to the original granter` covers it. A future minor version could swap the address stamp for a monotonic transfer counter to make invalidation permanent.

A deployed version of this design runs on BSC mainnet (`0x0fA3F984F7999d31C28055260637D1bCEA34919A`, UUPS, BscScan-verified). The reference implementation here is stripped of project-specific naming and dependencies.

### Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update

#### Checklist

- [x] Added tests
- [ ] Created showcase

### CODE REVIEW

File: contracts/extensions/VaultPermissionManagerReference.sol [hunk 1]
```diff
@@ -50,8 +57,8 @@ contract VaultPermissionManagerReference is
 
     struct Grant {
```
Comment (Rachestew-a11y) [inline contracts/extensions/VaultPermissionManagerReference.sol:57]: No permission

**Comment by christelbuchanan** [comment]
@tsu-j 

The `all-checks-passed` job is failing because one or more of the required upstream jobs did not pass. Based on the job logs, the condition check shows that at least one of these jobs failed:

- `compile`
- `test`
- `lint-and-format`

**Root Cause:** The workflow checks if any of the three required jobs have a `failure` status. The log shows the condition evaluated to true, triggering the exit code 1 and failing the job.

**Recommended Fix:**

To resolve this, you need to:

1. Identify which upstream job failed by checking the CI workflow run details for the `compile`, `test`, or l`int-and-format` jobs in PR #37.

2. **Once identified, fix the failing job** - common issues include:

- **Compilation errors:** Check contract syntax in `contracts/**/*.sol`
- **Test failures:** Run `npm test` locally to identify failing tests
- **Linting issues:** Run `npm run lint` and `npx prettier --check 'contracts/**/*.sol'` locally

3. **Update the workflow condition** (optional improvement): Consider excluding `security-analysis` from the required checks in the `all-checks-passed` job since it has `fail-on: none`:

```
YAML
all-checks-passed:
  needs: [compile, test, lint-and-format, security-analysis]
  if: always()
  runs-on: ubuntu-latest
  steps:
    - name: Check if all jobs passed
      run: |
        if [[ "${{ needs.compile.result }}" == "failure" || "${{ needs.test.result }}" == "failure" || "${{ needs.lint-and-format.result }}" == "failure" ]]; then
          echo "One or more required checks failed"
          exit 1
        fi
        echo "All checks passed!"
```
**Next Step:** Check the individual job results in the PR workflow run to determine which specific job failed and what error it produced.

File: contracts/extensions/VaultPermissionManagerReference.sol [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        uint256 currentEpoch
+    ) internal view returns (bool ok, uint256 expiry) {
+        Grant storage g = _grants[tokenId][vid][accessor];
+        if (!g.exists) return (false, 0);
+        if (uint8(g.level) < uint8(minLevel)) return (false, g.expiry);
+        if (g.expiry != 0 && block.timestamp > g.expiry) return (false, g.expiry);
+        // Auto-revoke: a grant is dead once the token's owner epoch has moved.
+        if (g.epoch != currentEpoch) return (false, g.expiry);
+        return (true, g.expiry);
+    }
```
[the remark this answers is not in the document]
Comment (christelbuchanan) [reply contracts/extensions/VaultPermissionManagerReference.sol:242]: @tsu-j please check

File: contracts/extensions/VaultPermissionManagerReference.sol [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+        uint256 currentEpoch = _syncEpoch(tokenId);
+        _grants[tokenId][vid][grantee] = Grant({
+            level: level,
+            expiry: expiry,
+            epoch: currentEpoch,
+            exists: true
+        });
+        emit PermissionGranted(tokenId, vid, grantee, level, expiry, description);
+    }
```
[the remark this answers is not in the document]
Comment (christelbuchanan) [reply contracts/extensions/VaultPermissionManagerReference.sol:148]: @tsu-j please check

File: contracts/mocks/MockLogicModule.sol [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            revertNext = false;
+            revert(bytes(revertMessage).length == 0 ? "MockLogicModule: forced revert" : revertMessage);
+        }
+        lastTokenId = tokenId;
+        lastAction = action;
+        lastData = data;
+        lastCaller = msg.sender;
+        return abi.encode(true);
+    }
+}
```
[the remark this answers is not in the document]
Comment (christelbuchanan) [reply contracts/mocks/MockLogicModule.sol]: @tsu-j please check

File: contracts/extensions/VaultPermissionManagerReference.sol [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        uint256 currentEpoch = _syncEpoch(tokenId);
+        (bool ok, ) = _check(tokenId, vid, msg.sender, PermissionLevel.WRITE, currentEpoch);
+        require(ok, "VPM: caller lacks WRITE");
+
+        (, , address logic, ) = IBAP578(bap578).agentStates(tokenId);
+        require(logic != address(0), "VPM: no logic bound");
+
+        emit ActionForwarded(tokenId, vid, msg.sender, action);
+        result = ILogicModuleForwardable(logic).handleAction(tokenId, action, data);
+    }
```
[the remark this answers is not in the document]
Comment (christelbuchanan) [reply contracts/extensions/VaultPermissionManagerReference.sol:178]: @tsu-j please check

File: contracts/extensions/VaultPermissionManagerReference.sol [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+    mapping(uint256 => mapping(bytes32 => mapping(address => Grant))) private _grants;
+
+    // tokenId => owner epoch. Advances every time the contract observes a new owner.
+    mapping(uint256 => uint256) private _ownerEpoch;
+
+    // tokenId => the owner address the contract last observed.
+    mapping(uint256 => address) private _lastSeenOwner;
+
+    // Reserved storage for future upgrades.
+    uint256[48] private __gap;
```
[the remark this answers is not in the document]
Comment (christelbuchanan) [reply contracts/extensions/VaultPermissionManagerReference.sol:79]: @tsu-j please check

[the remark this answers is not in the document]
Comment (christelbuchanan) [reply contracts/extensions/VaultPermissionManagerReference.sol:79]: @tsu-j please check

**Comment by christelbuchanan** [comment]
@tsu-j 

The `lint-and-format` job is failing due to **code style issues** detected by Prettier in 3 Solidity files. The job runs `npx prettier --check 'contracts/**/*.sol'` which fails when formatting violations are found.

**Failing Files:**
1. `contracts/extensions/IVaultPermissionManager.sol`
2. `contracts/extensions/VaultPermissionManagerReference.sol`
3. `contracts/mocks/MockLogicModule.sol`

**Root Cause:**
Prettier detected formatting inconsistencies in these files that don't match the project's code style configuration.

**Solution:**

Run Prettier with the `--write` flag to automatically fix the formatting issues:

```bash
npx prettier --write contracts/extensions/IVaultPermissionManager.sol
npx prettier --write contracts/extensions/VaultPermissionManagerReference.sol
npx prettier --write contracts/mocks/MockLogicModule.sol
```

Or fix all files at once:

```bash
npx prettier --write 'contracts/**/*.sol'
```

After running this command, commit the formatted files:

```bash
git add contracts/extensions/IVaultPermissionManager.sol contracts/extensions/VaultPermissionManagerReference.sol contracts/mocks/MockLogicModule.sol
git commit -m "fix: format Solidity files with Prettier"
git push
```

The `lint-and-format` job will then pass since the files will comply with the configured Prettier formatting rules.

**Comment by christelbuchanan** [comment]
@tsu-j 

This job is from the "Smart Contract Tests" workflow and is failing with the same code formatting issue as the previous job. **Prettier detected code style violations** in 3 Solidity files:

1. `contracts/extensions/IVaultPermissionManager.sol`
2. `contracts/extensions/VaultPermissionManagerReference.sol`
3. `contracts/mocks/MockLogicModule.sol`

**Root Cause:**
The workflow runs `npx prettier --check 'contracts/**/*.sol'` which validates code formatting but doesn't automatically fix it.

**Solution:**

Fix all formatting issues by running Prettier with the `--write` flag:

```bash
npx prettier --write contracts/extensions/IVaultPermissionManager.sol
npx prettier --write contracts/extensions/VaultPermissionManagerReference.sol
npx prettier --write contracts/mocks/MockLogicModule.sol
```

Or in a single command:

```bash
npx prettier --write 'contracts/**/*.sol'
```

Then commit and push the changes:

```bash
git add contracts/extensions/IVaultPermissionManager.sol contracts/extensions/VaultPermissionManagerReference.sol contracts/mocks/MockLogicModule.sol
git commit -m "fix: format Solidity files with Prettier"
git push
```

This will align the files with our Prettier configuration and allow the formatting check in the workflow to pass.

**Comment by BORT-AGENTS** [comment]
Hi Christel, thanks for the detailed review and for tagging each finding inline. Pushed b9557bf which addresses:

- Ran `prettier --write` on the five touched files; lint-and-format should pass on the next run.
- `revokePermission` now calls `_syncEpoch(tokenId)` before authorizing, matching the documented invariant Greptile flagged. Authorization behavior is unchanged (the view path already used `_currentEpoch`).
- Added a new test `lets an ADMIN-tier grantee call forwardHandleAction (ADMIN includes WRITE)` to close Greptile's test gap. Suite is now 38/38.
- `MockLogicModule` now inherits `ILogicModuleForwardable` (with `override` on `handleAction`), resolving the "Missing inheritance" alert.

The other Slither alerts you tagged are intentional. Short justification for each:

- **`__gap` not in mixedCase (Warning)**: `__gap` is the OpenZeppelin upgradeable convention for storage-reservation gaps. Renaming it would diverge from the standard pattern.
- **`__gap` unused state variable (Warning)**: a storage gap is unused on purpose; it reserves storage slots for future upgrades. Slither does not recognize the idiom.
- **`forwardHandleAction` ignores `agentStates` return values (Medium)**: `(, , address logic, ) = IBAP578(bap578).agentStates(tokenId);` is intentional tuple destructuring. Only `logicAddress` is needed; the other fields are explicitly discarded.
- **`grantPermission` and `_check` use `block.timestamp` (Low, x2)**: grant expiry is a timestamp, so the comparison is required.

If you would rather have the alerts clear on our side via `// slither-disable-next-line` directives instead of dismissing them, happy to push that as a follow-up.

Thanks for the review.
